numba/_helperlib.c

```diff
@@ -331,7 +331,10 @@ numba_adapt_ndarray(PyObject *obj, arystruct_t* arystruct) {
     npy_intp *p;
 
     if (!PyArray_Check(obj)) {
-        return -1;
+        obj = PyObject_CallMethod(obj, "__array__", NULL);
+        if (!obj) {
+            return -1;
+        }
```

Comment:
I think the code needs to run `PyArray_Check(obj)` again if `obj` is not `NULL`, just to make sure that the value returned from the `obj.__array__()` method call is actually an array.